File: src/nn/tools/histogram.r
library(readr)
library(ggplot2)

calculate_bins <- function(data) {
  data <- na.omit(data) # Remove NA values from the data
  
  q25 <- quantile(data, 0.25)
  q75 <- quantile(data, 0.75)
  iqr <- q75 - q25
  
  bin_width <- (2 * iqr) / (length(data)^(1/3))
  num_bins <- round((max(data) - min(data)) / bin_width)
  
  return(num_bins)
}

# Read the CSV file
df <- read_csv('molecules.csv')

# Extract the "Molecular weight (g mol-1)" column
emission_max <- df$`Emission max (nm)`

# Calculate the number of bins
num_bins <- calculate_bins(emission_max)

# Plot the histogram
plot_title <- 'Histogram of Emission Maxima'

ggplot(df, aes(x = emission_max)) + 
  geom_histogram(binwidth = num_bins, fill = "blue", color = "black") +
  labs(title = plot_title, x = 'Molecular weight [g/mol]', y = 'Frequency') + 
  theme_minimal() +
  theme(panel.grid.major = element_line(colour = "grey80")) +
  ggsave(filename = paste0(plot_title, ".png"))

File: src/nn/deprecate/tools/histogram.py
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

def calculate_bins(data):
    """
    Calculate the number of bins using the Freedman-Diaconis rule, accounting for NaN values.
    """
    data = data.dropna()  # Remove NaN values from the data
    q75, q25 = np.percentile(data, [75, 25])
    iqr = q75 - q25
    bin_width = (2 * iqr) / (len(data) ** (1/3))
    num_bins = int((max(data) - min(data)) / bin_width)
    return num_bins

# Read the CSV file
df = pd.read_csv('molecules.csv')

# Extract the "emission max [nm]" column
emission_max = df['Molecular weight (g mol-1)']

# Calculate the number of bins
num_bins = calculate_bins(emission_max)

# Plot the histogram
plt.hist(emission_max, bins=num_bins, edgecolor='black')

title = 'Histogram of Molecular Weight'
# Customize the plot
plt.xlabel('Molecular weight [g/mol]')
plt.ylabel('Frequency')
plt.title(title)
plt.grid(True)


plt.savefig(title)

# Display the plot
plt.show()
